src/sources/datadog/logs.rs

```diff
@@ -78,10 +78,16 @@ struct DatadogLogsSource {
     store_api_key: bool,
 }
 
+// https://github.com/DataDog/datadog-agent/blob/a33248c2bc125920a9577af1e16f12298875a4ad/pkg/logs/processor/json.go#L23-L49
 #[derive(Deserialize)]
 struct LogMsg {
-    pub message: String,
-    pub timestamp: i64,
+    pub message: Bytes,
+    pub status: Option<Bytes>,
+    pub timestamp: Option<i64>,
+    pub hostname: Option<Bytes>,
+    pub service: Option<Bytes>,
+    pub source: Option<Bytes>,
+    pub tags: Option<Bytes>,
```

Comment:
Does this still need a `HashMap` to pick up the remaining arbitrary key/values?

Reply:
My understanding from the agent code is that there wouldn't be further fields.

Reply:
Ah, I replied to a pre-edit. Well, that's a good question. Does this source support arbitrary clients that are allowed to send arbitrarily keyed messages or is it specifically to catch log payloads from the datadog agent?

Reply:
If we want the performance story here to be as good as possible we need to be specific about the keys that are inbound so we can map them to a struct without indirection into a collection but I'm also sympathetic to the idea that some users will end up retargetting their local vector install instead of the intake API, in which case this source needs to be as lax as that API about what inputs it accepts.

Reply:
Sorry, that edit could have probably been another comment. And yes, that is a good question 🙂

Historically, we've tried to keep sources/sinks compatible such that I could, for example, use the `datadog_logs` sink to push to a vector with a `datadog_logs` source. That alone seems like it'd require handling of the dynamic fields.

Having the structured fields you have here, along with a catch-all, might still be enough to see the performance benefit with logs coming from the datadog agent, depending on how serde handles the deserialization. However, if there is still a significant performance penalty, we could consider making it something you have to opt into.

Reply:
We do seem to suggest in the source documentation that it's a datadog agent doing the sending, so if we mean to be more expansive here it's worth documenting what the correct payload form is.

Reply:
Chatted with Ben about this and we think we could make the source specific to receiving from the agent with some additional notes:

* We rename this `datadog_logs` source to `datadog_agent`. This `datadog_agent` source will be the primary source for receiving all data from the agent (logs, metrics, traces). To start, we'd only support logs. Metrics is expected to come this quarter. The rename would also help make it clear that this is a "breaking change"; I think we could avoid adding the alias and have people explicitly opt over into it.
* We get a "contract" in-place with the Datadog agent team so that they notify us if the agent is going to modify the way it sends data to allow us to get a release out ahead of their changes to accommodate them.
* In the future, we may reintroduce a `datadog_logs` source that is intended to mimic the Datadog Logs ingest endpoint and would be fully compatible with the `dattadog_logs` sink.

Reply:
@jszwedko perfect. I think my appetite here is to get this PR merged and we follow-up with a rename / re-document issue since I want to be sure to isolate changes as much as possible for measurement purposes.

Reply:
👍 as it stands, this merge would be breaking. Would it be ok if I open a PR to move around `datadog_logs` first? I can do it tomorrow.

Reply:
Sure. I'm in no hurry and am happy to rebase this small change however.